Fix covariant returns with generic return types

janvorli opened the pull request:
When a covariant return type was an uninstantiated generic type, the
ClassLoader::IsCompatibleWith was not working properly. In debug builds,
it was asserting because there was no MethodTable for that type and in
release builds, it resulted in ExecutionEngineException or an internal
CLR error.

This change fixes it by using TypeHandle::CanCastTo instead of
MethodTable::CanCastTo and adds a regression test for two cases where
the problem was observed (Assembly.GetTypes() and creating an instance
of a type with a covariant return with a problematic kind of type).

Close #45037
Close #45082

janvorli committed:
Fix covariant returns with generic return types

When a covariant return type was an uninstantiated generic type, the
ClassLoader::IsCompatibleWith was not working properly. In debug builds,
it was asserting because there was no MethodTable for that type and in
release builds, it resulted in ExecutionEngineException or an internal
CLR error.

This change fixes it by using TypeHandle::CanCastTo instead of
MethodTable::CanCastTo and adds a regression test for two cases where
the problem was observed (Assembly.GetTypes() and creating an instance
of a type with a covariant return with a problematic kind of type).

benaadams commented:
Also fixes https://github.com/dotnet/runtime/issues/45082 ?

janvorli commented:
@benaadams I've just tried and it doesn't fix it. There is yet another issue hidden behind my fix (with my fix, it fails at a different place than before).

benaadams commented:
I did try changing the runtime to use covariant returns for RuntimeType; but it then crashes with both types of crossgen and jitdiff https://github.com/benaadams/runtime/commit/e19bcfd7a3af1f76273051bcc8dd1579be40446a

janvorli commented:
@benaadams can you please create an issue for that? I'll look into it after fixing the #45082.

benaadams commented:
Raised https://github.com/dotnet/runtime/issues/45356

janvorli committed:
Fix issue 45082 too

There were two issues. First, the
ClassLoader::ValidateMethodsWithCovariantReturnTypes was called before
typeHnd.DoFullyLoad and that resulted in an assert down the call chain
of TypeDesc::CanCastTo due to a wrong load level.
Second, the SigTypeContext generation for the current MD in the
ClassLoader::ValidateMethodsWithCovariantReturnTypes requires the same
change for class instantiation as the one that we had for the parent MD.

janvorli commented:
I've updated the PR so that it fixes the #45082 too and added a regression test for that one.

jkotas commented on src/coreclr/src/vm/clsload.cpp:
```diff
@@ -3531,16 +3531,15 @@ static void PushFinalLevels(TypeHandle typeHnd, ClassLoadLevel targetLevel, cons
 
+        if (!typeHnd.IsTypeDesc())
+        {
+            ClassLoader::ValidateMethodsWithCovariantReturnTypes(typeHnd.AsMethodTable());
```
DoFullyLoad marks the type as fully loaded. Another thread that tries to load the same type may see it as fully loaded and won't perform this validation.

I think this needs to be done before the type is marked as fully loaded.

janvorli replied:
Hmm, so I'll need to figure out how to fix the issue that was fixed by this change. The problem was that down the call chain of the ValidateMethodsWithCovariantReturnTypes, a load level assert fired.

davidwrighton replied:
This is the only concern I have with this change.

davidwrighton reviewed (changes requested):
Please fix the issue around DoFullyLoad

janvorli committed:
Fix the issue 45082 in a correct way

The call to ClassLoader::ValidateMethodsWithCovariantReturnTypes is now
in MethodTable::DoFullyLoad.
I have also added a test case that verifies a case that David Wrighton
has suggested offline, where there are 3 types... A, B and C.
C derives from B which derives from A. B has a bad override which
should produce an error.  Then, cause C to be fully loaded without
otherwise triggering a load of B.

janvorli commented:
The `Libraries Test Run release mono windows x64 Debug` leg keeps failing consistently in many libraries tests (even after closing / reopening the issue) and I cannot find previous failures of this leg. Moreover, this change doesn't touch anything from Mono stuff, so I am baffled what's going on. @jkotas or @marek-safar do you have any idea what can be going on?

jkotas commented:
@akoeplinger @safern Could you please disable the failing Mono legs or move them to the staging pipeline so that they do not cause all jobs to be red?

akoeplinger commented:
The leg was already disabled with https://github.com/dotnet/runtime/pull/45529 but you'd need to rebase/merge master to get this change in the PR. I'd just merge.

janvorli commented:
Thanks @akoeplinger!

janvorli commented:
/backport to release/5.0
